crates/sui-types/src/move_package.rs

```diff
@@ -94,11 +120,61 @@ impl MovePackage {
         version: SequenceNumber,
         module_map: &BTreeMap<String, Vec<u8>>,
         max_move_package_size: u64,
+        dependencies: &[MovePackage],
+    ) -> Result<Self, ExecutionError> {
+        // TODO: there may be a better Rust incantation for this but dealing with error mapping in
+        // closures is painful sometimes
+        let mut modules = vec![];
+        for bytes in module_map.values() {
+            modules.push(
+                CompiledModule::deserialize(bytes)
+                    .map_err(|_| ExecutionErrorKind::InvalidMoveModule)?,
+            );
+        }
```

Comment:
The rust incantation would probably be something like:

```rust
let modules = module_map
    .values()
    .map(CompiledModule::deserialize)
    .collect::<Result<Vec<_>, _>>()
    .map_err(|_| ExecutionErrorKind::InvalidModule)?;
```

But, having said that, I think the better way to go here is to have the type origin table generation on the code path where you already have access to the the deserialized `CompiledModule`s, and require that folks using the path using `new` have to supply the type origin table already.  I.e. I'm hoping that we can keep `MovePackage` initialisation to a handlful of paths:

- `MovePackage::new` which is just fed all the fields verbatim and just packages everything up (this will be useful for changing between types, like `SuiRawMovePackage` etc).
- `MovePackage::for_publish`, which is given everything it needs to compute the contents of a fresh package (a max size, an iterator of `CompiledModule`s, and an iterator of `&MovePackage` dependencies).  This will compute the linkage table from the dependencies (and do the consistency check with the linkage tables of its immediate dependencies), and compute the type origin table from the compiled modules.
- `MovePackage::for_upgrade` which is similar to the above, but also accepts a `predecessor: &MovePackage`.  It derives the package's version from the predecessors, and similar for its type origin table.
   - Note that this case is not so important initially, we can build this later, after the cut, the important one to work through is `for_publish`.
   - Originally I was going to suggest that this function also be responsible for the compatibility check with the predecessor package, but I think that's not a good idea -- the `Upgrade` command should do that, because it also needs to pay attention to the upgrade policy.

This set-up should make sure that the code paths that need to calculate the linkage and type origin tables get access to deserialized Compiled Modules (non-deserializable modules will have been weeded out, as is already the case for publish and upgrade), while the code paths that are just ferrying data around deal in bytes and precomputed tables.